Add bazel build infrastructure for Federation API

Martin-Idel opened the pull request:
This PR fixes the federation proto dependencies on internal tools and adds the corresponding infrastructure to build the protos (including golang definitions). 

In addition, a simple example shows how they can be used in another bazel project.

Martin-Idel committed:
Rename existing BUILD file

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add common directories to .gitignore

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Replace internal types by nmts types

We replace unknown types by nmts types.
Since there is no transceiver or bent pipe anymore,
we can also get rid of some message complexity.

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Use outernetcouncil nmts repository as git override

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add Bazel build for protos and go protos

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add examples how to use the API in a go project

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

electronjoe commented on BUILD:
```diff
@@ -1,4 +1,4 @@
-# Copyright 2024 Outernet Council Foundation
+# Copyright (c) Outernet Council and Contributors.
```
I don't have enough context to know if this change was requested, or is typical, etc - can you explain the Copyright changes here and elsewhere? Have they been applied uniformly to the Repo?

Martin-Idel replied:
My bad - I forgot to ask about this! The nmts repo uniformly uses this style and I accidentally copied the wrong version. Since there already is a contributors list here, maybe it makes sense to uniformly change to this version? 

However, for the sake of not changing anything that doesn't need to change, I reverted this change and changed all copyrights in all new files to match this version here.

electronjoe commented on BUILD:
```diff
@@ -12,46 +12,31 @@
-    ],
-    importpath = "aalyria.com/spacetime/api/fed/v1alpha",
-    proto = ":federation_proto",
+    importpath = "oc.com/smo/api/fed/v1alpha",
```
As you pointed @Martin-Idel, this import path (here and in proto) feels misaligned with typical practices.  Exploring [Google APIs](https://github.com/googleapis/googleapis/blob/master/google/api/servicecontrol/v2/service_controller.proto) repo, and some other modern proto repos, their approach seems to map to something like this for us:

From:
```protobuf
package oc.smo.api.fed.v1alpha;
option java_package = "com.oc.smo.api.fed.v1alpha";
option go_package = "github.com/outernetcouncil/federation/api/fed/v1alpha";
```

To:
```protobuf
package outernet.federation.v1alpha;
option go_package = "github.com/outernetcouncil/federation/gen/go/federation/v1alpha;federationpb";
option java_package = "com.outernet.federation.v1alpha";
```

Key improvements:

1. Package name:
   - Simplified to `outernet.federation.v1alpha`
   - More consistent with Google's patterns of organization.name.product.version
   - Removed unnecessary "api" segment since it's implied

2. go_package:
   - Added `gen/go` to indicate generated code
   - Added `pb` suffix to the package name to indicate it's protobuf generated code
   - Follows pattern seen in many Google APIs and other major gRPC projects

3. java_package:
   - Simplified to match the protobuf package structure more closely
   - Uses standard Java package naming conventions
   - Matches the overall namespace structure

electronjoe commented on MODULE.bazel:
```diff
@@ -0,0 +1,37 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module(name = "org_outernetcouncil_federation")
```
Here and elsewhere, what do you say to shortening the `org_outernetcouncil_federation` to at least `org_outernet_federation`?

I'm not highly opinionated either way, but the protobuf import paths for example I am advising just `outernet` for org, instead of the full `outernetcouncil`? Though again could be swayed for the longer option if we think it's important. Maybe @syotfs has an opion too.

Martin-Idel replied:
nmts is "org_outernetcouncil_nmts" so I chose `org_outernetcouncil_federation` for consistency.
My feeling was also that this is usually related to the package name so I would be fine with changing both as you suggest.

I added a commit to change it - but that's of course your call!

electronjoe commented on MODULE.bazel:
```diff
@@ -0,0 +1,37 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module(name = "org_outernetcouncil_federation")
```
**Module Declaration Enhancement**

I notice the `module()` declaration could be enhanced with version information and compatibility level. This helps with dependency resolution and maintenance:

```python
module(
    name = "org_outernetcouncil_federation",
    version = "1.0.0",
    compatibility_level = 1,
    # compatibility_level history
    # 1 -> version 1.x.x initial release
)
```

Adding `version` and `compatibility_level` provides several benefits:
- Makes the module's version explicit for dependent projects
- Enables Bazel's version selection features
- Helps prevent accidental breaking changes through compatibility levels

The compatibility level (starting at 1) indicates API compatibility - increment it when making breaking changes. This follows semantic versioning principles and helps downstream consumers manage updates.

For more details on module versioning best practices, see:
https://bazel.build/external/module#version

electronjoe commented on BUILD:
Lets rename this to BUILD.bazel I think that's more canonical?

electronjoe commented on examples/MODULE.bazel:
I don't think that a per-directory MODULE.baze file is best practice. Were you having some path mapping issues here? I think we should be able to make this work by changes to the root MODULE.bazel and dropping this sub-directory MODULE.bazel file...

Martin-Idel replied:
My idea was that the examples are a completely separate project so that they can be built independently to showcase how to build/use the real project. I took this over from the nmts repo, which also has an example folder with separate examples.

However, you are right: This is what makes `bazel build //...` fail, so it's probably best to get rid of it. I'll upload that shortly.

Martin-Idel replied:
nmts gets away with this structure probably because the examples only depend on nmts and have no other external dependencies so that the build definitions still work.

electronjoe reviewed (commented):
# Top level comments

Here and in all comments of this review, I'm obviously open to discussion - and if changes are agreed upon they can land either in this PR or in a subsequent PR.

## Build warnings

When I e.g. `bazel build //:federation_go_proto` I'm noticing a suggestion:

```shell
Imported, but reported as indirect dependencies by the extension:
    org_golang_google_genproto

Fix the use_repo calls by running 'bazel mod tidy'.
```

If I run `bazel mod tidy` it drops the `MODULE.bazel` use_repo for `org_golang_google_genproto` and breaks future bazel build attempts. Not sure what is going on here but possibly related to build errors below?

## Build Errors

I am of the habit of doing:

```shell
bazel build //...
```

This presently fails with:

```shell
bazel build //...
ERROR: no such package '@@[unknown repo 'org_golang_google_protobuf' requested from @@]//proto': The repository '@@[unknown repo 'org_golang_google_protobuf' requested from @@]' could not be resolved: No repository visible as '@org_golang_google_protobuf' from main repository
ERROR: /Users/smoeller/Documents/outernet/federation/examples/BUILD.bazel:17:10: no such package '@@[unknown repo 'org_golang_google_protobuf' requested from @@]//proto': The repository '@@[unknown repo 'org_golang_google_protobuf' requested from @@]' could not be resolved: No repository visible as '@org_golang_google_protobuf' from main repository and referenced by '//examples:federation_example'
Use --verbose_failures to see the command lines of failed build steps.
ERROR: Analysis of target '//examples:federation_example' failed; build aborted: Analysis failed
INFO: Elapsed time: 0.181s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
```

I see we don't use a go.mod yet. I attempted patching in to root BUILD file:

```/BUILD
load("@gazelle//:def.bzl", "gazelle")

gazelle(
    name = "gazelle",
    prefix = "github.com/outernetcouncil/federation",
)
```

and

```/MODULE.bazel
go_deps = use_extension("@gazelle//:extensions.bzl", "go_deps")
go_deps.from_file(go_mod = "//:go.mod")
```

```shell
rm examples/MODULE.bazel
go mod init github.com/outernetcouncil/federation
go mod tidy ## maybe an error to do this
bazel run @rules_go//go -- get google.golang.org/protobuf@latest
bazel run @rules_go//go -- mod tidy
bazel run //:gazelle
```

This did not work - nor did various iterations.

Other proto build targets work just fine (e.g. `//:federation_proto, //:federation_go_proto`). **I wonder if your `/examples/MODULE.bazel` use is confounding things here**.  See my comment in that file, I don't think this recursive MODULE.bazel file should be necessary and it could be mucking things up?

## Documentation

Doesn't have to be in this PR, but lets get some docs in place on how to build this for the first time for a new arrival (e.g. ?bazelisk?, some basic Bazel Build commands for this repo, a link to Bazel docs). This probably belongs at the bottom area of README.md down near the Detailed API Guide section?

Martin-Idel committed:
Get rid of nested MODULE file

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Rename BUILD file to canonical extension

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Rename packages for better naming convention

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Shorten name from outernetcouncil to outernet

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Add short developer documentation to README

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel committed:
Use consistent copyright schema for all new files

Signed-off-by: Martin Idel <martin.idel@tngtech.com>

Martin-Idel commented:
@electronjoe I updated the files with your suggestions.

You are right, the additional MODULES file confused the build so that "all targets" didn't work and since I only checked the targets individually, I didn't notice.

However, the build warnings still remain and I find both extremely puzzling, as there is a very direct dependency on both packages in the `go_proto_library` and `go_binary` target. If you have any insight here, I'd be glad, otherwise I can keep investigating (maybe we'll need a followup)

Other than this, I added a short description to the README under "CONTRIBUTING" for now. It was already there and empty and as building the project is usually the first step to contributing and whenever I see a section "CONTRIBUTING" or a file that is usually my starting point to see what I need to do to build a project.

I marked those comments as resolved, where I just took over your suggestions without change.

electronjoe reviewed (approved):
Nice work @Martin-Idel ! Thanks for getting this done.
